### PR TITLE
Fix for issue #21118: inconsistent behavior across callbacks

### PR DESCRIPTION
**Pull request by adar21a**
https://github.com/keras-team/keras/issues/21118

### CODE REVIEW

File: keras/src/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -197,6 +183,44 @@ def __init__(
                     f"filepath={self.filepath}"
                 )
 
+    def _set_monitor_op(self):
```
Comment (fchollet) [inline keras/src/callbacks/model_checkpoint.py:186]: Can we refactor this logic into a standalone function that we could reuse across all callbacks that need this functionality?

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: keras/src/callbacks/model_checkpoint.py (hunk 1 above)
Comment (adar21a) [reply keras/src/callbacks/model_checkpoint.py]: I added a `MonitorCallback` base class for all the callbacks that use this functionality (currently `EarlyStopping`, `ReduceLROnPlateau` and `ModelCheckpoint`). This ensures consistent behavior across callbacks and reduces code duplication.

File: keras/src/callbacks/monitor_callback.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
+import warnings
+
+from keras.src import ops
+from keras.src.api_export import keras_export
+from keras.src.callbacks.callback import Callback
+from keras.src.trainers import compile_utils
+
+
+@keras_export("keras.callbacks.MonitorCallback")
```
Comment (fchollet) [inline keras/src/callbacks/monitor_callback.py:9]: The refactoring into a shared base class makes sense, but please do not export it to the public API.

Comment (adar21a) [reply keras/src/callbacks/monitor_callback.py]: I removed it. However, I do think it could be useful as a public API — perhaps in a follow-up PR with some adjustments. It would be helpful for users who want to create custom callbacks that monitor a metric — for example, plotting something whenever the loss decreases.
